README.md
# Reaction plane fit

[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.1599239.svg)](https://doi.org/10.5281/zenodo.1599239)
[![Documentation Status](https://readthedocs.org/projects/reactionplanefit/badge/?version=latest)](https://reactionplanefit.readthedocs.io/en/latest/?badge=latest)
[![Build Status](https://travis-ci.com/raymondEhlers/reactionPlaneFit.svg?branch=master)](https://travis-ci.com/raymondEhlers/reactionPlaneFit)
[![codecov](https://codecov.io/gh/raymondEhlers/reactionPlaneFit/branch/master/graph/badge.svg)](https://codecov.io/gh/raymondEhlers/reactionPlaneFit)

Implements the reaction plane (RP) fit described in [Phys. Rev. C 93,
044915](https://journals.aps.org/prc/abstract/10.1103/PhysRevC.93.044915) (or on the
[arxiv](https://arxiv.org/abs/1509.04732)) to characterize and subtract background contributions to
correlation functions measured in heavy ion collisions. It allows for fits to the background dominated region,
as well as the option of fitting the RP inclusive signal dominated region or the RP dependent signal dominated
region. This package implements the fit for three orientations relative to the RP, with the possibility of
straightforward extension to other sets of orientations.

![Sample reaction plane fit](https://github.com/raymondEhlers/reactionPlaneFit/raw/master/docs/images/sampleSignalInclusiveRPF.png)

## Installation

This package requires python 3.6 and above. A few prerequisites are required which unfortunately cannot be
resolved solely by pip because of the packaging details of `probfit`.

```bash
$ pip install numpy cython
```

The package is available on [PyPI](https://pypi.org/project/reaction_plane_fit) and is available via pip.

```bash
$ pip install reaction_plane_fit
```

You may want to consider installing this to your user directory (`--user`) or better, within a virtual
environment.

# Usage

Performing a fit with this package only requires a few lines of code. Below is sufficient to define and run a
fit with some sample values:

```python
from reaction_plane_fit import three_orientations
# Define the fit object.
rp_fit = three_orientations.BackgroundFit(
    resolution_parameters = {"R22": 0.6, "R42": 0.3, "R62": 0.1, "R82": 0.1},
    use_log_likelihood = False,
    signal_region = (0, 0.6),
    background_region = (0.8, 1.2),
)
# Load or otherwise provide the relevant histograms here.
# The structure of this dictionary is important to ensure that the proper data ends up in the right place.
# Note that the data must already be projected into the background or signal dominated regions.
data = {"background": {"in_plane": ROOT.TH1.., "mid_plane": ROOT.TH1..., "out_of_plane": ROOT.TH1...}}
# Perform the actual fit.
success, _ = rp_fit.fit(data = data)
# Print the fit results
print("Fit result: {fit_result}".format(fit_result = rp_fit.fit_result))
```

For best fit results, the data should **not** be normalized by the number of triggers before performing the
fit. Afterwards, the fit needs to be scaled by the number of triggers in order to make it comparable to the
scaled data. Examples for fitting an inclusive reaction plane orientation signal alongside the background, or
for fitting only the background are both available in `reaction_plane_fit.example`. This module can also be
run directly in the terminal via:

```
$ python -m reaction_plane_fit.example [-b] [-i dataFilename]
```

If fit data is not specified, it will use some sample data. For further information, including all possible
fit function combinations, please see the [full documentation](https://reactionplanefit.readthedocs.io/en/latest/).

## But I don't like python!

You might not be so happy about using Python. That's okay - we can work around this using ROOT, although it
will be much more painful than using python directly. To do so, it should look something like (the code below
is **untested**, so it may need some minor modifications):

```c++
// Setup the input data.
std::map<std::string, std::map<std::string, TH1*>> inputData;
inputData["background"]["in_plane"] = my_in_plane_hist;
// Fill in the rest of your input data.
...
// Setup the fit objects.
TPython::Exec("from reaction_plane_fit import three_orientations");
TPython::Exec("rp_fit = three_orientations.BackgroundFit("
    "resolution_parameters = {'R22': 0.6, 'R42': 0.3, 'R62': 0.1, 'R82': 0.1},"
    "use_log_likelihood = False,"
    "signal_region = (0, 0.6),"
    "background_region = (0.8, 1.2),)"
);
// Perform the actual fit.
TPython::Exec("success, _ = rp_fit.fit(data = data)");
// Print the fit results.
TPython::Exec("print('Fit result: {fit_result}'.format(fit_result = rp_fit.fit_result))")
// Access values back in c++.
int chi_2 = double(TPython::Eval("rp_fit.fit_result.minimum_val"))
// It will require some attention to extract all of the relevant values.
// You can extract a number of types (see TPyResult), but it doens't appear that you can extract complex objects.
// So this could still be a somewhat painful process.
```

(Don't forgot that we're calling Minuit underneath, so this package should be fast enough for this fit).

# Fits implemented

There are three possible types of fits:

- Background dominated region only: This fits only regions on the near-side at large dEta which are dominated
  by background contributions. Called `BackgroundFit` in each set of orientations implementation.
- Inclusive signal region: This fits to the RP inclusive signal dominated region, as well as fitting the RP
  dependent background dominated regions. Called `InclusiveSignalFit` in each set of orientations implementation.
- RP dependent signal region: This fits to the RP dependent signal dominated region, as well as fitting the RP
  dependent background dominated regions. Called `SignalFit` in each set of orientations implementation.

## Three orientation (in-, mid-, and out-of-plane)

This package implements the fit for three orientations relative to the reaction plane:

- in-plane (0<|&Delta;&phi;|<&pi;/6)
- mid-plane (&pi;/6<|&Delta;&phi;|<&pi;/3)
- and out-of-plane (&pi;/3<|&Delta;&phi;|<&pi;/2)

These fits are implemented in the `reaction_plane_fit.three_orientations` module.

# Notes on fitting

This package makes it possible to use Minos rather than Hesse errors. Minos errors can be calculated when
Hesse errors become inaccurate (when the function around the minima is not approximately a hyperparabola),
but they can take much longer to calculate and cannot be described via a covariance matrix (which makes error
propagation much more difficult). If the Hesse and Minos errors are similar, then the function is well
approximately by a hyperparabola and you can safely use Hesse errors. For more, see the iminuit tutorials.

# Development

If developing the packaging, clone the repository and then install with

```bash
$ pip install -e .[dev,tests]
```

Note that python 3.6 and above is required because this package uses `dataclasses` (which has a python 3.6
backport), and it relies on dictionaries being ordered (which is true for `cpython` 3.6 and is required for
python 3.7 in general).

# Citation

Please cite the paper (Phys. Rev. C 93, 044915), as well as this implementation:

```
@misc{raymond_ehlers_2018_1599239,
  author       = {Raymond Ehlers},
  title        = {reactionPlaneFit - RPF implementation},
  month        = nov,
  year         = 2018,
  doi          = {10.5281/zenodo.1599239},
  url          = {https://doi.org/10.5281/zenodo.1599239}
}
```

# Acknowledgments

Code started from implementation work done by [M. Arratia](https://github.com/miguelignacio/BackgroundFit).
Thanks to C. Nattrass and J. Mazer for help and discussions.


CHANGELOG.md
# Reaction Plane Fit Changelog

Changelog based on the [format here](https://keepachangelog.com/en/1.0.0/).

## [4.0] - 21 October 2019

## Added

- Additional documentation. See: `c49fa753201250734f6dbc4bdce6a64959a277cb`.
- Integration test for differential signal fit. See: `efcfc40fdf662d5fb0297ed1d02c2f6b538e2485`.

## Changed

- Switch emphasis to fitting **before** normalizing by the number of triggers, including updating default
  arguments and documentation. See: `17a5d24397c313c3c5a39964e9a8bd471161153e`,
  `e04a1f3b1736c8baef41872680bd175f3dadd977`, `73f5f0d8dcbf7f86f1742e57279da43fd62faa1f`.
- When reading the RP fit object from YAML, setup the entire fit object. See: `1b38175ae421729a1933bae63bcc6652262927d3`.
- Improve Minuit fit strategy. See: `d8dfb49d0599f564597461270b73bb67d1818bdf`.
- Rename `_fit` to `cost_func` for clarity - it represents the cost function, not the fit function. And it's
  sometimes useful to access from outside the object (such as when calculating effective chi squared). See:
  `3dc750119be1f52f05134b11f24883b2e5c73b07`.
- Make the Minuit verbosity a user settable option. See: `22e67c85b630ef09a3105bcee4587e8d7091b293`.
- Allow a user to skip HESSE when necessary. This should be rare, but it is required at times. See:
  `98104086ca505a25778c75a0c63053085a836368`.
- Updated the signal differential fit procedure. The inclusive case is just treated as a background only fit.
  This is required because adding coupling gaussian distributions is not a straightforward process (unlike
  if they were uncoupled). Consequently, extracting the inclusive width relies on fitting to information already
  included in the RP fit (included via the differential signal fit components), which means that the covariance of
  the fit is neglected. Use care when using this fit approach! (A warning is thrown when using it). See:
  `7ba221d8effcda48d3a5245776aefcb971d7748e`.
- Sample data now acts as if it is the raw correlation rather than scaled by the number of triggers. See:
  `5c613d2f5d21882fe9d61d6d0067bd602cb74ff2`, `9484b457012f6647283e3fc1ebf43a719a0fd7e5`,
  `5c613d2f5d21882fe9d61d6d0067bd602cb74ff2`, and `7db8ea4e4201879c609c0cebeb211ea67570e7ba`.
- Improve testing procedure by loading stored fits from YAML rather than defining them by hand in the test
  file. This makes any future updates much easier. See: `5c613d2f5d21882fe9d61d6d0067bd602cb74ff2`.

## Fixed

- Log likelihood fits by updating the underlying `pachyderm.fit` package. See: `33bdcb41cd0f2dcc6e76f8760b8e5eddd24456dd`.
- Minor typos in code, documentation. See: `150afbba7f065f54895f1bd30409c66df14990b9` and
  `56af2759823cf2ff027c2b89d7047fdcb2bfc98f`.
- Create output directory when writing if necessary. See: `7ca7f01d12a897907427972c6a9f6402cba6c616`.
- Allow specification of `v_{1}` values. See: `ed037ba57f72dc24257dd6a2dbe1351e85e6c73d` and a fix in
  `63cb49906eabc115f58b6847de756346b61b08c3`.
- Updated typing. See: `db79b4134f57ff76727356654489b9a823cb6b67`.
- Integration tests run successfully on TravisCI again. See: `330c3bc6a7f90f1a5e0773c03170a8b27b8cf4ca`.

## [3.1] - 24 August 2019

## Changed

- Added a method to create a full set of fit components for each EP orientation, regardless of which were used
  for the fit. This allows for downstream code to not need to differentiate between the different cases for
  most tasks, easing the user experience. See: Nearly all of the commits between 3.0 and 3.1.
- Use the `pachyderm.fit.BaseFitResult` for components in order to avoid needing to calculate unnecessary
  values for component fit results. It's more natural. See: `bbbcac45c41490072f345b6aebbeb6f49d427a1d`.
- Updated the included test data for better test consistently with external sources. See:
  `0bf5e1ceaede9cd578a20c12db81455f38326a2d`.
- Updated pre-commit hooks. See: `6d780545b5547f42c7c3a684f11234d069973c10`.

## Fixed

- Warning about registering `pytest` marks. See: `8f6f5386fddea1e540a15de99ce8852dfc2319fd`.
- Removed final `probfit` dependence. See: `b94ea052246269e454f3df1c7afbb7ac96f35cec`.

## [3.0] - 8 August 2019

## Changed

- Move to using the pachyderm fit package for general fit functionality, cost functions, and fit results. This
  move was instigated by the recognition that `probfit` cost functions didn't actually work properly with
  pre-binned data. It also allows for the centralization of basic fit code to make it generally available.
  Some code from this package was refactored into pachyderm. See: `9e62f6777afed290709b74d4ee18284d61c2bab9`
  and `0f66cff8264ff06f5138d78ba0213f40a195dd5e`. Note that developed was handled in a private branch in this
  repository, so some older commits have references to a fit package in the RPF package before it was moved to
  pachyderm.
- No longer rely on `probfit`. All functionality and more is now provided in `pachyderm.fit`. Plus, the package
  wasn't super actively maintained. See: `9e62f6777afed290709b74d4ee18284d61c2bab9`.
- Updated pre-commit hooks. See: `03844495a12f4269b14dbd4972c09fb203a88c7c`.

## Fixed

- Log likelihood and chi squared cost functions properly take binned data into account. See:
  `0f66cff8264ff06f5138d78ba0213f40a195dd5e`.

## [2.1.2] - 4 May 2019

### Added

- Enabled strict `mypy` checks. See: `1585c7398201319bac49f6881f678be3f9582ee5`.

### Changed

- Improved typing. `341330a5e48f1526848de27ceaba4ee10499de67`.

## [2.1.1] - 3 May 2019

### Changed

- Refactor function error calculation so that it can used by other packages. See:
  `04dff6cc7cbd1fd4ad1f2f26ecc528dfeebbe3f7`.

### Fixed

- Enable fixing parameters via user arguments. See: `cf9de684245217d4c7583407940d5a3864b192b1`.
- Properly mark the package as typed. See: `fe43d5368bb8fdd1defa5dbff8a39ac45d2424a3`.
- Improved code formatting and typing information, and fixed typos.
- Fix `mypy` typing in third party packages by adding `zip_safe = False` in the `setup.py`. See:
  `79eb4e821bed8d5290c18c779c2bce6c8a71e5c0`.

## [2.1] - 29 March 2019


reaction_plane_fit/example.py
#!/usr/bin/env python

""" Example of how to use the Reaction Plane Fit package.

Provides example for fitting only the background, or fitting the inclusive signal.

.. code-author: Raymond Ehlers <raymond.ehlers@ycern.ch>, Yale University
"""

import argparse
import logging
import pkg_resources
from typing import Any, Dict, Tuple, Type, TYPE_CHECKING
import uproot

from reaction_plane_fit import fit
from reaction_plane_fit.fit import Data, InputData
from reaction_plane_fit.fit import FitArguments
from reaction_plane_fit import three_orientations
from reaction_plane_fit import plot

if TYPE_CHECKING:
    import iminuit  # noqa: F401

logger = logging.getLogger(__name__)

# Type helpers
FitReturnValues = Tuple[fit.ReactionPlaneFit, Dict[str, fit.FitComponent], Data, "iminuit.Minuit"]

def setup_data(input_filename: str) -> InputData:
    """ Setup the example input data.

    Read the data using uproot so we can avoid an explicit dependency on ROOT. Histograms
    are assumed to be named ``{region}_{orientation}`` (for example, "signalDominated_inclusive").

    Args:
        input_filename: Path to the input data to use.
    Returns:
        dict: Containing the input data.
    """
    data: Dict[str, Any] = {"signal": {}, "background": {}}
    with uproot.open(input_filename) as f:
        for region in ["signal", "background"]:
            for rp in ["inclusive", "in_plane", "mid_plane", "out_of_plane"]:
                data[region][rp] = f[f"{region}Dominated_{rp}"]

    return data

def run_fit(fit_object: Type[fit.ReactionPlaneFit],
            input_data: InputData,
            user_arguments: FitArguments) -> FitReturnValues:
    """ Driver function for performing the fit.

    Note:
        Here we just set the resolution parameters to 1 for convenience of the example, but they are an extremely
        important part of the fit!

    Args:
        fit_object: Fit object to be used.
        data: Input data for the fit, labeled as defined in ``setup_data()``.
        user_arguments: User arguments to override the arguments to the fit.
    Returns:
        tuple: (rp_fit, full_components, data, minuit), where rp_fit (fit.ReactionPlaneFit) is the reaction plane
            fit object from the fit, full_components is the full set of fit components (one for each RP orientation),
            data (dict) is the formated data dict used for the fit, and minuit is the minuit fit object.
    """
    # Define the fit.
    rp_fit = fit_object(
        resolution_parameters = {"R22": 1, "R42": 1, "R62": 1, "R82": 1},
        use_log_likelihood = False,
        signal_region = (0, 0.6),
        background_region = (0.8, 1.2),
    )

    # Perform the actual fit.
    success, data, minuit = rp_fit.fit(data = input_data, user_arguments = user_arguments)
    # Create the full set of fit components
    full_components = rp_fit.create_full_set_of_components(input_data = data)

    if success:
        logger.info(f"Fit was successful! Fit result: {rp_fit.fit_result}")

    return rp_fit, full_components, data, minuit

def run_background_fit(input_filename: str, user_arguments: FitArguments) -> FitReturnValues:
    """ Run the background example fit.

    Args:
        input_filename: Path to the input data to use.
        user_arguments: User arguments to override the arguments to the fit.
    Returns:
        tuple: (rp_fit, full_components, data, minuit), where rp_fit (fit.ReactionPlaneFit) is the reaction plane
            fit object from the fit, full_components is the full set of fit components (one for each RP orientation),
            data (dict) is the formated data dict used for the fit, and minuit is the minuit fit object.
    """
    # Grab the input data.
    input_data = setup_data(input_filename)
    return_values = run_fit(
        fit_object = three_orientations.BackgroundFit,
        input_data = input_data, user_arguments = user_arguments
    )
    return return_values

def run_inclusive_signal_fit(input_filename: str, user_arguments: FitArguments) -> FitReturnValues:
    """ Run the inclusive signal example fit.

    Args:
        input_filename: Path to the input data to use.
        user_arguments: User arguments to override the arguments to the fit.
    Returns:
        tuple: (rp_fit, full_components, data, minuit), where rp_fit (fit.ReactionPlaneFit) is the reaction plane
            fit object from the fit, full_components is the full set of fit components (one for each RP orientation),
            data (dict) is the formated data dict used for the fit, and minuit is the minuit fit object.
    """
    input_data = setup_data(input_filename)
    return_values = run_fit(
        fit_object = three_orientations.InclusiveSignalFit,
        input_data = input_data, user_arguments = user_arguments
    )
    return return_values

def run_differential_signal_fit(input_filename: str, user_arguments: FitArguments) -> FitReturnValues:
    """ Run the differential signal example fit.

    Args:
        input_filename: Path to the input data to use.
        user_arguments: User arguments to override the arguments to the fit.
    Returns:
        tuple: (rp_fit, full_components, data, minuit), where rp_fit (fit.ReactionPlaneFit) is the reaction plane
            fit object from the fit, full_components is the full set of fit components (one for each RP orientation),
            data (dict) is the formated data dict used for the fit, and minuit is the minuit fit object.
    """
    input_data = setup_data(input_filename)
    return_values = run_fit(
        fit_object = three_orientations.SignalFit,
        input_data = input_data, user_arguments = user_arguments
    )
    return return_values

if __name__ == "__main__":  # pragma: nocover
    """ Allow direction execution of this module.

    The user can specify the input filename and which type of fit to perform. However, the names of the input
    histograms must be as specified in ``setup_data(...)``.
    """
    # Setup logging
    logging.basicConfig(level = logging.INFO)
    # Quiet down the matplotlib logging regardless of the base logging level.
    logging.getLogger("matplotlib").setLevel(logging.INFO)
    # Setup parser
    parser = argparse.ArgumentParser(
        description = "Example Reaction Plane Fit using signal and background dominated sample data."
    )

    sample_data_filename = pkg_resources.resource_filename("reaction_plane_fit.sample_data", "three_orientations.root")
    # Set the input filename
    parser.add_argument("-i", "--inputData", metavar = "filename",
                        type = str, default = sample_data_filename,
                        help="Path to input data")
    # Set the fit type (defaults to differential signal fit)
    parser.add_argument("-d", "--differentialSignal",
                        action = "store_true",
                        help = "Fit the differential signal regions.")
    parser.add_argument("-b", "--backgroundOnly",
                        action = "store_true",
                        help = "Only fit the background.")
    # Parse arguments
    args = parser.parse_args()

    # Execute the selected function
    func = run_inclusive_signal_fit
    if args.differentialSignal:
        func = run_differential_signal_fit
    if args.backgroundOnly:
        func = run_background_fit
    rp_fit, _, data, _ = func(
        input_filename = args.inputData,
        user_arguments = {},
    )
    # Save to YAML.
    rp_fit.write_fit_results("example.yaml")

    # Determine fit label
    fit_label_map = {
        three_orientations.BackgroundFit: "Standard RP fit",
        three_orientations.InclusiveSignalFit: "Inclusive signal RP fit",
        three_orientations.SignalFit: "Differential signal RP fit",
    }
    fit_label = fit_label_map[type(rp_fit)]

    # Draw the plots so that they will be saved out.
    plot.draw_fit(rp_fit = rp_fit, data = data, fit_label = fit_label, filename = "example_fit.png")
    plot.draw_residual(rp_fit = rp_fit, data = data, fit_label = fit_label, filename = "example_residual.png")

